sdk/authentication.py

```diff
@@ -0,0 +1,146 @@
+            raise TokenProviderError(err)
+
+    @staticmethod
+    def _extract_expiry(token_response: dict[str, Any]) -> datetime | None:
+        """Extracts the expiry datetime from the token response.
+
+        Assumes that the actual token provided is correct, but might be in an
+        unexpected format
+
+        Args:
+            token_response: A valid token response dictionary.
+
+        Returns:
+            The time the token will expire, or None if it can't be calculated
+        """
+        expires_in = token_response.get("expires_in", None)
+        if expires_in:
+            return datetime.now(tz=timezone.utc) + timedelta(seconds=float(expires_in))
+
+        try:
+            # We assume the token is valid, but might not be in an expected format
+            token = token_response.get("access_token")
+
+            if isinstance(token, str) and "." in token:
+                token_timestamp = decode(
+                    token,
+                    algorithms=["HS256"],
+                    options={"verify_signature": False},
+                ).get("exp")
+                if token_timestamp:
+                    return datetime.fromtimestamp(token_timestamp, tz=timezone.utc)
+
+        except DecodeError:
+            # The token is not in a format that could be parsed as a JWT
+            pass
+        return None
+
+
+class Auth0TokenProvider(TokenProvider):
+    def __init__(self, username: str, password: str, auth0: Auth0Conf):
+        super().__init__(username, password, auth0)
+
+        # Makes a call in order to check the credentials at creation
+        self._query_token()
+
+    def _query_token(self) -> dict[str, Any]:
+        token = GetToken(self.auth0.domain)
+
+        # No client secret required for this Application since
+        # "Token Endpoint Authentication Method" set to None
+        validated_token: dict[str, Any] = token.login(
+            client_id=self.auth0.public_client_id,
+            client_secret="",
+            username=self.username,
+            password=self.password,
+            audience=self.auth0.audience,
+            scope="openid profile email",
+            realm=self.auth0.realm,
+            grant_type="http://auth0.com/oauth/grant-type/password-realm",
+        )
+        return validated_token
+
+
+class FakeAuth0GoodAuthentication(TokenProvider):
+    def _query_token(self) -> dict[str, Any]:
+        return {
+            "access_token": "some_token",
+            "id_token": "id_token",
+            "scope": "openid profile email",
+            "expires_in": 86400,
+            "token_type": "Bearer",
+        }
+
+
+class FakeAuth0BadAuthentication(TokenProvider):
+    def __init__(self, *args: tuple, **kwags: dict):
+        """The arguments are not important.
+        What's important is that the init raise the error below.
+        """
+        self._query_token()
+
+    def _query_token(self) -> dict[str, Any]:
+        raise ValueError()
```

Comment:
This should be the same error as the one returned by auth0 when the credentials are incorrect and we should test manually that the resulting error message is clear for the user (and if not catch the Auth0 error and raise a custom one).

Reply:
This was originally a `Auth0Error`, but I noticed that I didn't want to test for a Auth0Error in Mocked Test, in order to be able to differentiate if the query was mocked or not. If the query is not Mocked, how can I be sure that the error is from the Auth0Mock ? Or the Auth0 real client ?

I saw this because I placed a bad test inside `TestAuthInvalidClient`, but it made a real request, that was green in the test because I was checking for Auth0Error. If I had been checking for `ValueError`, I would have gotten a `Auth0Error` in the test that wasn't mocked (and therefore a red light, what I wanted).

What do you think ?

Reply:
You can test directly that your mock was called using `assert_called_once` or equivalent instead of returning a different error. Theoretically speaking I'd rather raise the same errors as the object we are mocking for peace of mind :)

Reply:
done